jit/src/lib.rs

```diff
@@ -96,23 +111,76 @@ pub struct CompiledCode {
 }
 
 impl CompiledCode {
-    pub fn invoke(&self) -> Option<AbiValue> {
+    pub fn args_builder(&self) -> ArgsBuilder<'_> {
+        ArgsBuilder::new(self)
+    }
+
+    pub fn invoke<'a>(&self, args: &Args<'a>) -> Option<AbiValue> {
+        debug_assert_eq!(self as *const _, args.code as *const _);
         let cif = self.sig.to_cif();
         unsafe {
             let value = cif.call::<UnTypedAbiValue>(
                 libffi::middle::CodePtr::from_ptr(self.code as *const _),
-                &[],
+                &args.cif_args,
             );
             self.sig.ret.as_ref().map(|ty| value.to_typed(ty))
         }
     }
 }
 
+struct JitSig {
+    args: Vec<JitType>,
+    ret: Option<JitType>,
+}
+
+impl JitSig {
+    fn to_cif(&self) -> libffi::middle::Cif {
+        let ret = match self.ret {
+            Some(ref ty) => ty.to_libffi(),
+            None => libffi::middle::Type::void(),
+        };
+        libffi::middle::Cif::new(self.args.iter().map(JitType::to_libffi), ret)
+    }
+}
+
+#[derive(Copy, Clone, PartialEq)]
```

Comment:
The reason I didn't make this Copy is because I was thinking of eventually having composite types, like `List(Box<JitType>)` or `Tuple(Vec<JitType>)` or maybe some sort of product or sum type.